### PR TITLE
chore(otel): add microbenchmarks

### PR DESCRIPTION
**Pull request by mabdinur**
Benchmark creating, ending, and adding attributes to otel spans.

## Datadog API Span Benchmarks (against [1.x](https://github.com/DataDog/dd-trace-py/commit/d381a15e5fdf58be894099794103514b789dbd5f))
.....................
span-start: Mean +- std dev: 12.4 ms +- 1.3 ms (per 1000 spans)
.....................
span-add-tags: Mean +- std dev: 62.9 ms +- 2.0 ms (per 1000 spans)
.....................
span-add-metrics: Mean +- std dev: 39.1 ms +- 1.8 ms (per 1000 spans)
.....................
span-start-finish: Mean +- std dev: 11.0 ms +- 0.1 ms (per 1000 spans)

## Datadog OpenTelemetry API Span Benchmarks (against [1.x](https://github.com/DataDog/dd-trace-py/commit/d381a15e5fdf58be894099794103514b789dbd5f))
.....................
otelspan-start: Mean +- std dev: 14.9 ms +- 0.8 ms (per 1000 spans)
.....................
otelspan-add-tags: Mean +- std dev: 84.8 ms +- 3.4 ms (per 1000 spans)
.....................
otelspan-add-metrics: Mean +- std dev: 92.7 ms +- 2.4 ms (per 1000 spans)
.....................
otelspan-start-finish: Mean +- std dev: 13.6 ms +- 0.4 ms (per 1000 spans)

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: benchmarks/bm/utils.py [hunk 1]
```diff
@@ -7,8 +7,11 @@
 
 from six import BytesIO
 
+
```
Comment (mabdinur) [inline benchmarks/bm/utils.py:10]:
```suggestion
```

File: benchmarks/otel_span/scenario.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+from ddtrace import tracer
+from ddtrace.opentelemetry import TracerProvider
+
+
+set_tracer_provider(TracerProvider())
+os.environ["OTEL_PYTHON_CONTEXT"] = "ddcontextvars_context"
+otel_tracer = get_tracer(__name__)
+
+# To reduce noise avoid setting finished spans to the agent. Instead write encoded spans to stdout
```
Comment (mabdinur) [inline benchmarks/otel_span/scenario.py:16]:
```suggestion
# To reduce noise avoid setting finished spans to the agent.
```

File: benchmarks/otel_tracer/scenario.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+set_tracer_provider(TracerProvider())
+os.environ["OTEL_PYTHON_CONTEXT"] = "ddcontextvars_context"
+otel_tracer = get_tracer(__name__)
+
+
+utils.drop_traces(tracer)
+
+
+class OtelSpan(bm.Scenario):
```
Comment (mabdinur) [inline benchmarks/otel_tracer/scenario.py:20]:
```suggestion
class OtelTracer(bm.Scenario):
```

**Comment by mabdinur** [comment]
Once this PR is merged we will need to update the relenv-microbenchmarking-platform to run these new benchmarks on all PRs: [relenv-microbenchmarking-platform/dd-trace-py/steps/post-pr-comment.sh](https://github.com/DataDog/relenv-microbenchmarking-platform/blob/dd-trace-py/steps/post-pr-comment.sh)

File: .gitlab/benchmarks.yml [hunk 4]
```diff
@@ -51,6 +51,11 @@ benchmark-span:
   variables:
     SCENARIO: "span"
 
+benchmark-span:
```
Comment (mabdinur) [inline .gitlab/benchmarks.yml:54]:
```suggestion
benchmark-otel-tracer:
```

File: .gitlab/benchmarks.yml [hunk 5]
```diff
@@ -51,6 +51,11 @@ benchmark-span:
   variables:
     SCENARIO: "span"
 
+benchmark-otel-tracer:
+  extends: .benchmarks
+  variables:
+    SCENARIO: "otel_tracer"
+
```
Comment (mabdinur) [inline .gitlab/benchmarks.yml:58]: Failing against 1.x, I may need to introduce the gitlab benchmarks in a follow up PR

```suggestion
```

**Review by emmettbutler — APPROVED** [review]

File: benchmarks/span/scenario.py [hunk 6]
```diff
@@ -24,7 +29,7 @@ def run(self):
         def _(loops):
             for _ in range(loops):
                 for i in range(self.nspans):
-                    s = utils.gen_span("test." + str(i))
+                    s = tracer.trace("test." + str(i))
```
Comment (mabdinur) [inline benchmarks/span/scenario.py:32]:
```suggestion
                    # create a root span
                    s = tracer.start_span("test." + str(i))
```

File: benchmarks/otel_span/scenario.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        metrics = utils.gen_metrics(self)
+        setmetrics = len(metrics) > 0
+
+        # run scenario to include finishing spans
+        finishspan = self.finishspan
+
+        def _(loops):
+            for _ in range(loops):
+                for i in range(self.nspans):
+                    s = otel_tracer.start_span("test." + str(i))
```
Comment (mabdinur) [inline benchmarks/otel_span/scenario.py:42]:
```suggestion
                   # create a root span
                    s = otel_tracer.start_span("test." + str(i))
```

**Comment by avara1986** [comment]
@mabdinur IIUC, we need to add this scenario to https://github.com/DataDog/dd-trace-py/blob/1.x/.gitlab/benchmarks.yml, isn't it? Thanks!

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**